store/hash.go

```diff
@@ -0,0 +1,187 @@
+// Copyright 2016 The rkt Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package store
+
+import (
+	"crypto/sha256"
+	"crypto/sha512"
+	"fmt"
+	"hash"
+	"reflect"
+	"strings"
+)
+
+type HashAlgorithm interface {
+	// ValidateHashString checks that a given input hash string is valid
+	ValidateHashString(hashStr string) error
+	// NormalizeHashString, given an input hash string, returns the
+	//normalized hash string depending on the hash algorithm implementation
+	NormalizeHashString(hashStr string) (string, error)
+	// HashToHashString takes a hash.Hash and returns a hash string. The
+	// hash.Hash must be of the supported type.
+	HashToHashString(h hash.Hash) (string, error)
+	// NewHash returns a new hash.Hash of the supported hashAlgorithm type
+	NewHash() hash.Hash
+}
+
+// SHA512HashAlgorithm is a hash algorithm that uses sha512 sums
+// To ameliorate excessively long hash string, hash strings use only the first
+// half of a sha512 rather than the entire hash.
+type SHA512HashAlgorithm struct {
+	hashPrefix              string
+	lenHash                 int
+	lenSum                  int
+	lenHashString           int
+	lenNormalizedSum        int
+	lenNormalizedHashString int
+	minlenHashString        int
+}
+
+func NewSHA512HashAlgorithm() *SHA512HashAlgorithm {
+	hashPrefix := "sha512"
+	lenHash := sha512.Size  // raw byte size
+	lenSum := (lenHash) * 2 // in hex characters
+	lenHashString := len(hashPrefix) + 1 + lenSum
+	lenNormalizedSum := (lenHash / 2) * 2 // half length, in hex characters
+	lenNormalizedHashString := len(hashPrefix) + 1 + lenNormalizedSum
+	minlenHashString := len(hashPrefix) + 1 + 2 // at least sha512-aa
+
+	return &SHA512HashAlgorithm{
+		hashPrefix:              hashPrefix,
+		lenHash:                 lenHash,
+		lenSum:                  lenSum,
+		lenHashString:           lenHashString,
+		lenNormalizedSum:        lenNormalizedSum,
+		lenNormalizedHashString: lenNormalizedHashString,
+		minlenHashString:        minlenHashString,
+	}
+}
+
+func (ha *SHA512HashAlgorithm) ValidateHashString(hashStr string) error {
```

Comment:
Also godoc with small mention what we are actually validating and potential failure modes.